ballerina/websocket_utils.bal

```diff
@@ -0,0 +1,119 @@
+    RootFieldVisitor rootFieldVisitor = new (node);
+    parser:FieldNode fieldNode = <parser:FieldNode>rootFieldVisitor.getRootFieldNode();
+    stream<any, error?>|json sourceStream = getSubscriptionResponse(engine, schema, context, fieldNode);
+    if sourceStream is stream<any, error?> {
+        record {|any value;|}|error? next = sourceStream.next();
+        while next !is error? {
+            OutputObject outputObj = engine.getResult(node, context, next.value);
+            ResponseFormatter responseFormatter = new (schema);
+            OutputObject coercedOutputObject = responseFormatter.getCoercedOutputObject(outputObj, node);
+            if coercedOutputObject.hasKey(DATA_FIELD) || coercedOutputObject.hasKey(ERRORS_FIELD) {
+                check sendWebSocketResponse(caller, customHeaders, WS_NEXT, coercedOutputObject.toJson(), connectionId);
+            }
+            next = sourceStream.next();
+        }
+        if next is error {
+            check sendWebSocketResponse(caller, customHeaders, WS_ERROR, next.message(), connectionId);
+        } else {
+            check sendWebSocketResponse(caller, customHeaders, WS_COMPLETE, null, connectionId);
+        }
+    } else {
+        check sendWebSocketResponse(caller, customHeaders, WS_ERROR, sourceStream, connectionId);
+    }
+}
+
+isolated function validateSubscriptionPayload(json|WSPayload data, Engine engine) returns parser:OperationNode|json {
+    json|error payload = data is WSPayload ? data?.payload : data;
+    if payload is error {
+        return {errors: [{message: "Invalid format in WebSocket payload"}]};
+    }
+    json|error document = payload.query;
+    if document is error {
+        return {errors: [{message: "Query not found"}]};
+    }
+    if document !is string {
+        return {errors: [{message: "Invalid format in request parameter: query"}]};
+    }
+    if document == "" {
+        return {errors: [{message: "Empty query is found"}]};
+    }
+    json|error variables = payload.variables;
+    variables = variables is error && variables.message() == "{ballerina/lang.map}KeyNotFound" ? () : variables;
+    if variables is error || variables !is map<json>? {
+        return {errors: [{message: "Invalid format in request parameter: variables"}]};
+    }
+    parser:OperationNode|OutputObject validation = engine.validate(document, getOperationName(payload), variables);
+    if validation is parser:OperationNode {
+        return validation;
+    }
+    return validation.toJson();
+}
+
+isolated function getSubscriptionResponse(Engine engine, __Schema schema, Context context,
+                                          parser:FieldNode node) returns stream<any, error?>|json {
+    ExecutorVisitor executor = new(engine, schema, context, {});
+    any|error result = getSubscriptionResult(executor, node);
+    if result !is stream<any, error?> {
+        return {errors: [{message: "Error/s occurred in the subscription resolver"}]};
+    }
```

Comment:
@Nuvindu Can't the subscription resolvers return `error` values? If the resolvers can produce errors, the error message will not be appeared in the response due to this logic. The error message is always overridden with the `Error/s occurred in the subscription resolver`.

Reply:
Yes. I should bind the error message to the Websocket payload. Missed it.